internal/resources/cluster/data_source_cluster.go

```diff
@@ -25,116 +25,162 @@ import (
 
 func DataSourceTMCCluster() *schema.Resource {
 	return &schema.Resource{
-		ReadContext: dataSourceTMCClusterRead,
+		ReadContext: schema.ReadContextFunc(dataSourceTMCClusterRead(withTerraformConstructType(constructTypeDataSource))),
```

Comment:
what is the advantage of this approach ?

Reply:
@shreyassreenivas we are using this for secrets as well (for some terraform construct type behaviours), it is already used for a different reason in policies, to keep things consistent. Also, I think functional parameters are always memory safe.

If you feel strongly that we should take the context-based approach do let us know for Kubernetes secret as well, since we have already referred to policy resource and made changes according to that.

Reply:
The source of invocation is not a functional parameter and something which can be referred from contexts we should leverage that. Contexts are meant to be used for these use cases.

I believe the present code compromises on the readability of the handlers and I would like to keep it plain and simple. 
I feel the source of the function call can be inferred from the context

Moreover, functional parameters should be used when we have multiple options being passed.

Reply:
Yeah, I agree, @Shubbhang351 let's take this MR as a reference for the k8's secret resource.